hoomd/md/PairModulator.h

```diff
@@ -0,0 +1,351 @@
+}
+
+        HOSTDEVICE shape_type() { }
+
+#ifndef __HIPCC__
+
+        shape_type(pybind11::object shape_params, bool managed)
+            : envelope(pybind11::len(shape_params), managed)
+            {
+                pybind11::list shape_param = shape_params;
+                for (size_t i = 0; i < pybind11::len(shape_param); i++)
+                    {
+
+                        envelope[int(i)] = typename DirectionalEnvelope::shape_type(shape_param[i]);
+                    }
+            }
+
+        pybind11::object toPython()
+            {
+                pybind11::list envelope_py;
+                for (size_t i = 0; i < envelope.size();  i++)
+                    {
+                        envelope_py.append(envelope[int(i)].toPython());
+                    }
+                return envelope_py;
+            }
+#endif
+
+#ifdef ENABLE_HIP
+        //! Attach managed memory to CUDA stream
+        void set_memory_hint() const {
+            envelope.set_memory_hint();
+}
+#endif
+        ManagedArray<typename DirectionalEnvelope::shape_type> envelope;
+    };
+
+    //! Constructs the pair potential evaluator
+    /*!
+      \param _dr Displacement vector pointing from particle rj to ri
+      \param _rcutsq Squared distance at which the potential is set to 0
+      \param _quat_i Quaternion of the ith particle
+      \param _quat_j Quaternion of the jth particle
+      \param _params Per type pair parameters of the potential
+    */
+    DEVICE PairModulator( const Scalar3& _dr,
+                          const Scalar4& _quat_i,
+                          const Scalar4& _quat_j,
+                          const Scalar _rcutsq,
+                          const param_type& _params)
+        : dr(_dr),
+          rsq(_dr.x*_dr.x + _dr.y*_dr.y + _dr.z*_dr.z),
+          rcutsq(_rcutsq),
+          quat_i(_quat_i),
+          quat_j(_quat_j),
+          params(_params)
+        { }
+
+    //! If diameter is used
+    DEVICE static bool needsDiameter()
+        {
+            return (PairEvaluator::needsDiameter() || DirectionalEnvelope::needsDiameter());
+        }
+
+    //! Accept the optional diameter values
+    /*!
+      \param di Diameter of particle i
+      \param dj Diameter of particle j
+    */
+    DEVICE void setDiameter(Scalar di, Scalar dj)
+        {
+        }
+
+    //! Whether pair potential requires charges
+    DEVICE static bool needsCharge()
+        {
+            return (PairEvaluator::needsCharge() || DirectionalEnvelope::needsCharge());
+        }
+
+    //! Accept the optional charge values
+    /*!
+      \param qi Charge of particle i
+      \param qj Charge of particle j
+    */
+    DEVICE void setCharge(Scalar qi, Scalar qj)
+        {
+            // store qi and qj for later
+            // if (PairEvaluator::needsCharge())
+            //     pairEval.setCharge(qi, qj);
+            // if (DirectionalEnvelope::needsCharge())
+            //     envelEval.setCharge(qi, qj);
```

Comment:
```suggestion
        m_charge_i = qi;
        m_charge_j = qj;
```
Need to make member variables that store the charges and tags. A followup comment in `evaluate` uses these.